index.rst

```diff
@@ -1,61 +1,173 @@
-   See the README at https://github.com/lsst-sqre/lsst-technote-bootstrap or
-   this repo's README for more info.
+Broadly speaking, the Alert Database will have three uses:
 
-   Feel free to delete this instructional comment.
+1. It will serve as an authoritative historical record, archiving all the alert data that the project has sent out through alert streams.
+2. It will also permit science users to retrieve individual alert packets using an identifier that they got from somewhere else (for example, from a circular, or a query to the PPDB).
+3. Finally, it will assist users who are developing software by providing large batches of alert data which can be used to generate a simulated stream, test online algorithms, and evaluate alert filtering techniques.
 
-:tocdepth: 1
+Each of these use cases has design consequences.
 
-.. Please do not modify tocdepth; will be fixed when a new Sphinx theme is shipped.
+Use cases: a historical record
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-.. sectnum::
+Because the Alert Database acts as a historical record, it must keep all data in perpetuity.
+This amounts to a few terabytes per year (see "Data Sizing Calculations" below).
+
+To be a good historical record, we need to take data durability seriously.
+All writes to the Database from the Alert Production system must be fully stored before they are acknowledged, and we must be able to back up the Alert Database to protect from data loss.
+
+The only operations that should be permitted are reads and writes; updates and deletes should not be possible in the normal course of events.
+They might be necessary for corrective action if there has been a serious bug, but they certainly should not be exposed to ordinary users.
+
+Use cases: queryable DB
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The Alert Database needs to support "needle in a haystack" queries based on specific a single alert identifier.
+In particular, `get(alert_id)`` ought to retrieve the single alert packet associated with a specific alert_id.
+
+This should generally respond within a few seconds.
+
+In general, we can expect recent data to be queried more often than old data for this purpose, but everything needs to be available.
+
+Use cases: resource for developers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Users who want to test algorithms, generate simulated streams, or evaluate filters will be interested in retrieving large batches of alert packets.
+They will not need to retrieve these batches very often, and it's entirely acceptable for their requests to take a long time (even days) to fulfill.
+
+Data Sizing Calculations
+------------------------
+
+DMTN-102 :cite:`DMTN-102` provides estimates for the total volume of alerts received.
+We expect a maximum of about 10 million alerts per night, each about 80 kilobytes.
+LSE-163 :cite:`LSE-163` projects about 300 nights of observations per year.
+
+Combining these figures, we project 80 gigabytes of alert data per night, and 2.4 terabytes per year.
```

Comment:
This math is wrong by a factor of 10: 80KB * 10 million alerts = 800 gigabytes.

cc @ktlim and @ebellm

Reply:
Doh.  Can't believe I missed that.  Also, 800*300 = 240 TB, no?  That starts to make space-saving look better.  What about compression?

Reply:
Gah, yes, more factors! Thanks, yes, that's significant.

Reply:
> What about compression?

So, Avro files support a bunch of compression schemes: deflate, gzip, zstd, snappy, and bz2. By "support" I mean that they store the compression scheme in a plaintext header, and then the data portion of an Avro file is the chunk that's compressed.

I haven't measured compression ratios on LSST alert packets. It's worth testing, based on this volume of data.

One option to boost compression ratios might be to try to reuse dictionaries across multiple alert packets. We could do that by storing many alert packets in one file, or by being very clever and careful with how we initialize our Avro file writers to reuse a compressor. But first - I should figure out whether naive compression does a good job.

---

Also, Avro Object Container format would mean including a 50KB schema in the plaintext header. That's not compressed by Avro. If we used one alert per file, then we'd have one schema per alert. Storing 50KB * 10 million alerts * 300 nights = 150 TB per year, which is just the same JSON document repeated 3 billion times, does not seem wise.

Reply:
I measured compression of the [sample batch of alerts based on HITS](https://github.com/lsst-dm/sample_alert_info/):


| codec | file with 1 alert | file with 50 alerts | file with 217 alerts |
| --------- | --------------------| ------------------------|----------------- |
| null | 93 KB | 2.352 MB | 22.426 MB |
| deflate | 67 KB | 1.113 MB | 15.454 MB |
| snappy | 71 KB | 1.250 MB | 17.592 MB |
| ztsd | 68 KB | 1.113 MB | 15.464 MB |
| bz2 | 69 KB | 1.117 MB | 15.583 MB |

(the file with 217 alerts may compress worse because it includes some alerts with very large cutouts, which probably compress poorly)

So we're basically looking at a 20-60% reduction in volume by using compression, something in that range.

Reply:
I have made some major revisions to try to deal with this:
 - gzip every alert
 - store schemas separately from alert packet data
 - drop the idea of storing a duplicate copy for bulk export - bulk export will just be slow, thats OK